apache/metron
# METRON-1707 Port Profiler to Spark [Feature Branch]

This ports the Profiler to run in Spark.

This is a pull request against the `METRON-1699-create-batch-profiler` feature branch.

This is dependent on the follow PRs.  By filtering on the last commit, this PR can be reviewed before the others are reviewed and merged.
- [x] #1145 
- [x] #1146 
- [x] #1148 
- [x] #1147

This does not include a mechanism to launch the Batch Profiler yet.  A follow-on PR will provide a means to launch the Batch Profiler from the command line.

## Testing

### Test Existing Functionality

1. Launch the development environment.
1. Ensure alerts are created in the Alerts UI.
1. Ensure the Service Check in Ambari passes.
1. Follow the instructions in the Profiler README to create a basic profile in the REPL.
1. Follow the instructions in the Profiler README to create a simple profile using the Profiler topology in Storm.

### Test New Functionality

1. Build and install Metron.
    ```
    mvn clean install -DskipTests -T2C
    ```

1. Edit `metron-analytics/metron-profiler-spark/src/test/resources/log4j.properties` and set `log4j.logger.org.apache.metron.profiler.spark=DEBUG`.

1. Run the `BatchProfilerIntegrationTest` either in your IDE or using Maven like so.
    ```
    cd metron-analytics/metron-profiler-spark/
    mvn -Dtest=BatchProfilerIntegrationTest test    
    ```

1. The output should look something like the following which shows that 2 profile measurements were written by the Batch Profiler.
```
Running org.apache.metron.profiler.spark.BatchProfilerIntegrationTest
2018-08-13 09:09:53 DEBUG BatchProfiler:70 - Building 1 profile(s)
2018-08-13 09:09:53 DEBUG BatchProfiler:102 - Loading telemetry from 'src/test/resources/telemetry.json'
2018-08-13 09:09:56 DEBUG BatchProfiler:109 - Found 13710 telemetry record(s)
ANTLR Tool version 4.5 used for code generation does not match the current runtime version 4.7ANTLR Runtime version 4.5 used for parser compilation does not match the current runtime version 4.7ANTLR Tool version 4.5 used for code generation does not match the current runtime version 4.7ANTLR Runtime version 4.5 used for parser compilation does not match the current runtime version 4.72018-08-13 09:09:58 DEBUG BatchProfiler:127 - Generated 13710 message route(s)
2018-08-13 09:10:00 DEBUG ProfileBuilderFunction:84 - Building a profile for group 'profile1-192.168.66.1-1704628' from 4017 message(s)
2018-08-13 09:10:00 DEBUG ProfileBuilderFunction:98 - Profile measurement created; profile=profile1, entity=192.168.66.1, period=1704628, value=4017
2018-08-13 09:10:00 DEBUG ProfileBuilderFunction:84 - Building a profile for group 'profile1-192.168.138.158-1704628' from 9693 message(s)
2018-08-13 09:10:00 DEBUG ProfileBuilderFunction:98 - Profile measurement created; profile=profile1, entity=192.168.138.158, period=1704628, value=9693
2018-08-13 09:10:00 DEBUG BatchProfiler:167 - Produced 2 profile measurement(s)
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:123 - 1 profile measurement(s) written to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:123 - 1 profile measurement(s) written to HBase
2018-08-13 09:10:00 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:111 - About to write profile measurement(s) to HBase
2018-08-13 09:10:01 DEBUG HBaseWriterFunction:123 - 0 profile measurement(s) written to HBase
2018-08-13 09:10:01 DEBUG BatchProfiler:185 - 2 profile measurement(s) written to HBase
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.967 sec - in org.apache.metron.profiler.spark.BatchProfilerIntegrationTest
```

## Pull Request Checklist
